File: Day2/solp2.py
def is_safe_sequence(levels):
    # Check if sequence is strictly increasing or decreasing with differences 1-3
    if len(levels) <= 1:
        return True
        
    is_increasing = levels[1] > levels[0]
    
    for i in range(len(levels) - 1):
        diff = abs(levels[i] - levels[i + 1])
        if diff < 1 or diff > 3:
            return False
        if is_increasing and levels[i] >= levels[i + 1]:
            return False
        if not is_increasing and levels[i] <= levels[i + 1]:
            return False
    
    return True

def is_safe_with_dampener(levels):
    # First check if safe without removing any number
    if is_safe_sequence(levels):
        return True
    
    # Try removing each number one at a time
    for i in range(len(levels)):
        new_levels = levels[:i] + levels[i+1:]
        if is_safe_sequence(new_levels):
            return True
    
    return False

# Count safe reports
safe_count = 0
with open('inputs.txt', 'r') as f:
    for line in f:
        if line.strip():
            levels = list(map(int, line.split()))
            if is_safe_with_dampener(levels):
                safe_count += 1

print(f"Number of safe reports with Problem Dampener: {safe_count}")

File: Day2/solp2.swift
import Foundation

func isSafeSequence(_ levels: [Int]) -> Bool {
    // Check if sequence is strictly increasing or decreasing with differences 1-3
    guard levels.count > 1 else {
        return true
    }

    let isIncreasing = levels[1] > levels[0]

    for i in 0..<levels.count - 1 {
        let diff = abs(levels[i] - levels[i + 1])
        if diff < 1 || diff > 3 {
            return false
        }
        if isIncreasing && levels[i] >= levels[i + 1] {
            return false
        }
        if !isIncreasing && levels[i] <= levels[i + 1] {
            return false
        }
    }

    return true
}

func isSafeWithDampener(_ levels: [Int]) -> Bool {
    // First check if safe without removing any number
    if isSafeSequence(levels) {
        return true
    }

    // Try removing each number one at a time
    for i in 0..<levels.count {
        var newLevels = levels
        newLevels.remove(at: i)
        if isSafeSequence(newLevels) {
            return true
        }
    }

    return false
}

// Count safe reports
var safeCount = 0

// Get the current file's directory
let currentFilePath = #file
let currentDirectory = URL(fileURLWithPath: currentFilePath).deletingLastPathComponent().path

// Construct path to inputs.txt in the same directory
let inputPath = currentDirectory + "/inputs.txt"

if let fileContents = try? String(contentsOfFile: inputPath, encoding: .utf8) {
    let lines = fileContents.components(separatedBy: .newlines)

    for line in lines {
        let trimmed = line.trimmingCharacters(in: .whitespaces)
        if !trimmed.isEmpty {
            let levels = trimmed.split(separator: " ").compactMap { Int($0) }
            if isSafeWithDampener(levels) {
                safeCount += 1
            }
        }
    }
}

print("Number of safe reports with Problem Dampener: \(safeCount)")
